### PR TITLE
FloatMathDSL, Shader Dev Window, Color Slider Breakout

### PR DESCRIPTION
**Pull request by Tyler-IN**
add shader dev window

break out color Slider

add ColorRect control

add FloatMathDsl & flavors

### CODE REVIEW

File: Robust.Client/UserInterface/Control.cs [hunk 1]
```diff
@@ -474,6 +474,61 @@ public void AddChild(Control child)
             ChildAdded(child);
         }
 
+
+        /// <summary>
+        ///     Make the provided control a parent of this control.
```
Comment (gradientvera) [inline Robust.Client/UserInterface/Control.cs]: Shouldn't this be the opposite?

File: Robust.Shared/Utility/FloatMathDsl.cs [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+using Robust.Shared.Interfaces.Timing;
+using Robust.Shared.IoC;
+using Robust.Shared.Random;
+using StringToExpression;
+
+namespace Robust.Shared.Utility
+{
+
+    /// <summary>
+    /// Represents a language to that handles basic mathematics out-of-game.
```
Comment (gradientvera) [inline Robust.Shared/Utility/FloatMathDsl.cs]:
```suggestion
    /// Represents a language that handles basic mathematics out-of-game.
```

File: Robust.Shared/Utility/InGameFloatMathDsl.cs [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+using Robust.Shared.Interfaces.Random;
+using Robust.Shared.Interfaces.Timing;
+using Robust.Shared.IoC;
+using Robust.Shared.Random;
+
+namespace Robust.Shared.Utility
+{
+
+    /// <summary>
+    /// Represents a language to that handles basic mathematics in-game.
```
Comment (gradientvera) [inline Robust.Shared/Utility/InGameFloatMathDsl.cs]:
```suggestion
    /// Represents a language that handles basic mathematics in-game.
```

File: Robust.Client/UserInterface/Control.cs (hunk 1 above)
Comment (Tyler-IN) [reply Robust.Client/UserInterface/Control.cs]:
```suggestion
        ///     Make the provided control a child of this control.
```
Was copy pasta.

File: Robust.Client/UserInterface/Control.Styling.cs [hunk 4]
```diff
@@ -25,7 +25,7 @@ public partial class Control
             }
         }
 
-        public ICollection<string> StyleClasses { get; }
+        public ICollection<string> StyleClasses { get; set; }
```
Comment (PJB3005) [inline Robust.Client/UserInterface/Control.Styling.cs:28]: This should absolutely not be public set. It gets set to a special collection implementation that correctly updates the internal set and queues for style updates.

File: Robust.Client/UserInterface/Control.cs [hunk 5]
```diff
@@ -474,6 +474,61 @@ public void AddChild(Control child)
...
+        /// <exception cref="InvalidOperationException">
+        ///     Thrown if we already have a component with the same name,
+        ///     or the provided component is still parented to a different control.
+        /// </exception>
+        /// <exception cref="ArgumentNullException">
+        ///     <paramref name="child" /> is <c>null</c>.
+        /// </exception>
+        protected void InsertChild(int index, Control child)
+        {
+            DebugTools.Assert(!Disposed, "Control has been disposed.");
```
Comment (PJB3005) [inline Robust.Client/UserInterface/Control.cs:492]: Unify these precondition checks with `AddChild()`.

File: Robust.Client/UserInterface/Controls/ColorRect.cs [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+            set;
+        }
+
+        protected internal override void Draw(DrawingHandleScreen handle)
+        {
+            base.Draw(handle);
+
+            handle.UseShader(Shader);
+
+            if (!TryGetStyleProperty("color", out Color color))
```
Comment (PJB3005) [inline Robust.Client/UserInterface/Controls/ColorRect.cs:47]: You've got this backwards. Local properties (like `Color` here) should override the style properties. Not the other way around.

File: Robust.Client/UserInterface/Controls/ColorRect.cs [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+            {
+                color = _color ?? DefaultColor;
+            }
+
+            handle.DrawRect(UIBox2.FromDimensions(Vector2.Zero, PixelSize), color);
```
Comment (PJB3005) [inline Robust.Client/UserInterface/Controls/ColorRect.cs:52]:
```suggestion
            handle.DrawRect(PixelSizeBox, color);
```

File: Robust.Client/UserInterface/Controls/ColorSlider.cs [hunk 8]
```diff
@@ -0,0 +1,95 @@
+﻿#nullable enable
```
Comment (PJB3005) [inline Robust.Client/UserInterface/Controls/ColorSlider.cs:1]: `Robust.Client` is already nullable enabled. This isn't necessary.

File: Robust.Client/UserInterface/Controls/ColorSlider.cs [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+                    _colorValue = (byte) result;
+                    _slider.Value = result;
+                    _ignoreEvents = false;
+
+                    OnValueChanged?.Invoke();
+                }
+            };
+        }
+
+        public ColorSlider(string styleClass) : this()
```
Comment (PJB3005) [inline Robust.Client/UserInterface/Controls/ColorSlider.cs:90]: This kind of styling should instead be done with a CSS child selector:

```css
ColorSlider .fooClass > Slider {
	/* Properties */
}
```

**Review by PJB3005 — CHANGES_REQUESTED** [review]

**Comment by AJCM-git** [comment]
superseded by #1407

**Pull request closed without merging**